projects/nginx/fuzz/http_request_fuzzer.c

```diff
@@ -0,0 +1,321 @@
+  c->read->ready=1;
+  c->recv = reply_recv_handler;
+  return in->next;
+}
+
+
+// Create a base state for Nginx without starting the server
+int InitializeNginx(void) {
+  ngx_log_t *log;
+  ngx_cycle_t init_cycle;
+
+  if( access("nginx.sock", F_OK ) != -1 ) {
+    remove("nginx.sock");
+  }
+
+  ngx_debug_init();
+  ngx_strerror_init();
+  ngx_time_init();
+#if (NGX_PCRE)
+  ngx_regex_init();
+#endif
+
+  // Just output logs to stderr
+  ngx_log.file = &ngx_log_file;
+  ngx_log.log_level = NGX_LOG_EMERG;
+  ngx_log_file.fd = ngx_stderr;
+  log = &ngx_log;
+
+#if (NGX_OPENSSL)
+  ngx_ssl_init(log);
+#endif
+  ngx_memzero(&init_cycle, sizeof(ngx_cycle_t));
+  init_cycle.log = log;
+  ngx_cycle = &init_cycle;
+
+  init_cycle.pool = ngx_create_pool(1024, log);
+
+  // Set custom argv/argc
+  my_argv[0] = arg1;
+  my_argv[1] = NULL;
+  ngx_argv = ngx_os_argv = my_argv;
+  ngx_argc = 0;
+
+  // Weird trick to free a leaking buffer always caught by ASAN
+  // We basically let ngx overwrite the environment variable, free the leak and
+  // restore the environment as before.
+  char *env_before = environ[0];
+  environ[0] = my_argv[0] + 1;
+  ngx_os_init(log);
+  free(environ[0]);
+  environ[0] = env_before;
+
+  ngx_crc32_table_init();
+  ngx_preinit_modules();
+
+  FILE *fptr = fopen(config_file, "w");
+  fprintf(fptr, configuration);
+  fclose(fptr);
+  init_cycle.conf_file.len = strlen(config_file);
+  init_cycle.conf_file.data = (unsigned char *) config_file;
+
+  cycle = ngx_init_cycle(&init_cycle);
+
+  ngx_os_status(cycle->log);
+  ngx_cycle = cycle;
+
+  ngx_event_actions.add = add_event;
+  ngx_event_actions.init = init_event;
+  ngx_io.send_chain = send_chain;
+  ngx_event_flags =1;
+  ngx_event_timer_init(cycle->log);
+  return 0;
+}
+
+
+void *invalid_call(void) {
+  return NULL;
+}
+
+
+int LLVMFuzzerInitialize(int *argc, char ***argv){
+  return InitializeNginx();
+}
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t data_len) {
+  // have two free connections, one for client, one for upstream
+  ngx_event_t read_event1 = {};
+  ngx_event_t write_event1 = {};
+  ngx_connection_t local1 = {};
+  ngx_event_t read_event2 = {};
+  ngx_event_t write_event2 = {};
+  ngx_connection_t local2 = {};
+  ngx_connection_t *c;
+  ngx_listening_t *ls;
+
+  req_reply=NULL;
+  upstream=NULL;
+  cln_added=0;
+
+  request.data = data;
+  request.data_len = data_len/2;
+  reply.data = data+data_len/2;
+  reply.data_len = data_len - data_len/2;
+
+  // Use listening entry created from configuration
+  ls = (ngx_listening_t*)ngx_cycle->listening.elts;
+
+  // Fake event ready for dispatch on read
+  local1.read = &read_event1;
+  local1.write = &write_event1;
+  local2.read = &read_event2;
+  local2.write = &write_event2;
+  local2.send_chain = send_chain;
+
+  // Create fake free connection to feed the http handler
+  ngx_cycle->free_connections = &local1;
+  local1.data = &local2;
+
+  ngx_cycle->free_connection_n = 2;
```

Comment:
it would be great if you could also add quick comments about value `2` here and value `255` on the line 285 -- why these values were selected?